Avoid Attribute.GetCustomAttributes() returning null for open generic type

madelson opened the pull request:
Fix #64335

madelson committed:
Avoid Attribute.GetCustomAttributes() returning null for open generic type.

Fix #64335

madelson commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:
```diff
@@ -1126,16 +1122,13 @@ internal static bool IsAttributeDefined(RuntimeModule decoratedModule, int decor
         }
 
         private static object[] GetCustomAttributes(
-            RuntimeModule decoratedModule, int decoratedMetadataToken, int pcaCount, RuntimeType? attributeFilterType)
```
[Every caller of this method](https://source.dot.net/#System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs,4abc6a05a3812753,references) checks `Debug.Assert(caType is not null)` first, so this shouldn't have to be nullable.

madelson committed:
Fix GenericAttributeTests to reflect #64335

Previously, this test asserted that GetCustomAttributes on an open generic
type would return null. Now we return empty instead.

madelson committed:
Skip new attribute tests on Mono.

Joe4evr commented on src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs:
```diff
@@ -438,6 +438,9 @@ private static Attribute[] CreateAttributeArrayHelper(Type elementType, int elem
         }
+
+        private static Attribute[] ToAttributeArray(object[] getCustomAttributesResult) =>
+            getCustomAttributesResult as Attribute[] ?? Array.Empty<Attribute>();
```
This was also an option I considered, but I wasn't sure if there's a code path that would return a populated array (somehow) and now those elements would be discarded.

buyaa-n replied:
I agree with @Joe4evr, a null value only returned when a generic attribute requested, adding null check for all overlaods wold affect perf. I think the the code section @Joe4evr [pointed out in his comment](https://github.com/dotnet/runtime/issues/64335#issuecomment-1026193422) is where we should lookup for a fix
> We could fall back to an empty Attribute[] (what I went with here).

That is the expected behavior when there is no any attribute found

madelson replied:
@buyaa-n can you clarify what you're suggesting here?

> I think the the code section @Joe4evr https://github.com/dotnet/runtime/issues/64335#issuecomment-1026193422 is where we should lookup for a fix

I did change those lines (see below). However, that doesn't solve the general problem that having an `as` cast here will result in returning null for any case where the underlying `GetCustomAttributes()` function doesn't return an `Attribute[]`. It didn't feel right to leave that in place.

> adding null check for all overlaods wold affect perf

Are you saying that adding `?? Array.Empty<Attribute>()` to handle the "cast failed" case will be a perf issue? If so, how do you think we should handle the case where the cast fails?

> I wasn't sure if there's a code path that would return a populated array (somehow) and now those elements would be discarded.

As I understand it, what @Joe4evr is saying here is that we could end up handling values like `new object[] { new SomeAttribute() }` or `new string[] { "a" }` or even `new object[] { new SomeAttribute(), "a" }`. What is the preferred behavior in such cases? Before this change we would return `null`. After this change we will return `Array.Empty<Attribute>()`. However, we could do something else.

buyaa-n replied:
> I did change those lines (see below). However, that doesn't solve the general problem that having an as cast here will result in returning null for any case where the underlying GetCustomAttributes() function doesn't return an Attribute[]. It didn't feel right to leave that in place.

I mean it was not returning (and not expected to return) null before this open generic attributes case introduced, and as you fixed that open generic case below i don't think we need this defensive mechanism, even if by the existing code it could return null in case an object array returned from the internal calls. Though I understand your reasoning, but as @Joe4evr mentioned now that object array creation will be discarded, probably we should change the object array returning logic instead

madelson replied:
> i don't think we need this defensive mechanism, even if by the existing code it could return null in case an object array returned from the internal calls

Ok thanks I understand. Do you think we should return to the `as` cast from before or should we do a "strict" cast `(Attribute[])member.GetCustomAttributes(...)`?

Also, I'm curious to get your thoughts on the desired behavior for the `CustomFieldInfo` test case I added:
| Case | `memberInfo.GetCustomAttributes()` return value | Old behavior | Initial PR behavior |
| ---- | ---- | ---- | ---- |
| 1 | `null` | `return null` | `return Array.Empty<Attribute>()` |
| 2 | `new object[0]` |  `return null` | `return Array.Empty<Attribute>()` |
| 3 | `new string[] { "a" }` | `return null` | `return Array.Empty<Attribute>()` |
| 4 | `new object[] { new GenericAttribute<string>, "b" }` | `return null` | `return Array.Empty<Attribute>()` |

> now that object array creation will be discarded, probably we should change the object array returning logic instead

As I understand it, this comment is referring to making further changes in the `CustomAttribute` class in `RuntimeCustomAttributeData.cs` to stop returning `object[]` at all. Currently, the cases where we still return an array with a non-attribute element type are:
(1) The passed in "attribute" type is a value type
(2) The passed in "attribute" type is a reference type that is not assignable to `Attribute`.

For example, here's some code that shows how this behavior is externally visible:
```C#
        Console.WriteLine(typeof(int).GetCustomAttributes(typeof(string), false).GetType()); // System.String[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(List<>), false).GetType()); // System.Object[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(int), false).GetType()); // System.Object[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(object), false).GetType()); // System.Object[]
	Console.WriteLine(typeof(int).GetCustomAttributes(typeof(object), false).Length); // 3 (actually populated Object[]!)

        // also works with TypeBuilder
	var assembly = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("asm"), AssemblyBuilderAccess.RunAndCollect);
	var module = assembly.DefineDynamicModule("mod");
	var type = module.DefineType("t");
	type.CreateType();
	Console.WriteLine(type.GetCustomAttributes(typeof(string), false).GetType()); // System.String[]
	Console.WriteLine(type.GetCustomAttributes(typeof(List<>), false).GetType()); // System.Object[]
	Console.WriteLine(type.GetCustomAttributes(typeof(int), false).GetType()); // System.Object[]
```

The reason that this behavior leaks out is that `RuntimeType` and `TypeBuilder` call into `CustomAttribute` directly without validating that the passed in type is an `Attribute` type for this one overload of `GetCustomAttributes`. This frankly feels like a bug to me, although the ability to request `object` does actually work.

Thoughts on what the best path forward is here and what impliciations there are for how we should handle unexpected return values in `Attribute.cs`?

jkotas replied:
Note that it is fine for attributes to not inherit from `System.Attribute`. It is not possible to create such attributes in C#, but it is possible in IL and other languages, in addition to the Reflection.Emit case that you have run into.

madelson replied:
@jkotas which framework APIs are supposed to work with such attributes?

Today, it seems that most do not work because you run into a cast to Attribute somewhere along the way.

jkotas replied:
APIs that return `object` and `object[]`: `MemberInfo.GetCustomAttributes`, etc.

buyaa-n replied:
> Ok thanks I understand. Do you think we should return to the as cast from before or should we do a "strict" cast (Attribute[])member.GetCustomAttributes(...)?

Personally I would use strict casting and let it throw, and it will be same as `GetCustomAttributes` for `Assemply` which throws InvalidCastException in such situation

buyaa-n commented on src/libraries/System.Runtime/tests/System/Attributes.cs:
```diff
@@ -533,6 +533,35 @@ public static void NegTest8()
+
+        // reproduces https://github.com/dotnet/runtime/issues/64335
+        [Fact]
+        [SkipOnMono("Mono throws on open generic type passed to GetCustomAttributes")]
```
I beleive we should fix it in mono too, there is a [CoreLib project](https://github.com/dotnet/runtime/tree/main/src/mono/System.Private.CoreLib) within mono source

buyaa-n replied:
Looks like [generic attributes are not supported](https://github.com/dotnet/runtime/issues/56887) on mono at all, so let keep skipping mono test.

buyaa-n commented on src/libraries/System.Runtime/tests/System/Attributes.cs:
```diff
@@ -533,6 +533,35 @@ public static void NegTest8()
+        [SkipOnMono("Mono throws on open generic type passed to GetCustomAttributes")]
+        public static void GetCustomAttributesDoesNotReturnNullForOpenGenericType()
+        {
+            Attribute[] openGenericAttributes = Attribute.GetCustomAttributes(typeof(HasGenericAttribute), typeof(GenericAttribute<>));
```
Please add test/fix for all member types including assembly, i got `InvalidCastException` when trying to get generic attribute from assembly
```log
System.InvalidCastException
  HResult=0x80004002
  Message=Unable to cast object of type 'System.Object[]' to type 'System.Attribute[]'.
  Source=System.Private.CoreLib
  StackTrace:
   at System.Runtime.CompilerServices.CastHelpers.ChkCastAny(Void* toTypeHnd, Object obj)
   at System.Attribute.GetCustomAttributes(Assembly element, Type attributeType, Boolean inherit)
   at System.Attribute.GetCustomAttributes(Assembly element, Type attributeType)
   at ConsoleApp13.Program.Main(String[] args) in \ConsoleApp13\Program.cs:line 22
```

buyaa-n commented on src/libraries/System.Runtime/tests/System/Attributes.cs:
```diff
@@ -533,6 +533,35 @@ public static void NegTest8()
 
         }
+
+        // reproduces https://github.com/dotnet/runtime/issues/64335
+        [Fact]
+        [SkipOnMono("Mono throws on open generic type passed to GetCustomAttributes")]
```
```suggestion
        [Fact]
        [ActiveIssue("https://github.com/dotnet/runtime/issues/56887)", TestRuntimes.Mono)]
```

jkotas commented:
Could you please also fix it for NativeAOT here: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Attribute.CoreRT.cs#L149

madelson commented:
> Could you please also fix it for NativeAOT here: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Attribute.CoreRT.cs#L149

@jkotas is your proposal to return `Array.Empty<Attribute>()` in that case instead of `null`? Do we still want to use the exception-based control flow?

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs:
```diff
@@ -438,6 +438,9 @@ private static Attribute[] CreateAttributeArrayHelper(Type elementType, int elem
             return (Attribute[])Array.CreateInstance(elementType, elementCount);
         }
+
+        private static Attribute[] ToAttributeArray(object[] getCustomAttributesResult) =>
```
Is there an actual real situation where `Array.Empty<Attribute>();` gets exercised now?

All places where this is called specify that the attribute has to inherit from `Attribute` or check that the requested attribute type inherits from `Attribute`. It means that we will never end up with an array that is not assignable to `Attribute[]` here, unless there is a bug. I think replacing the `!` with throwing cast would work just fine, and it would be simpler than this fix.

madelson replied:
> replacing the ! with throwing cast would work just fine, and it would be simpler than this fix.

I've done this.

> Is there an actual real situation where Array.Empty<Attribute>(); gets exercised now?

Not sure whether this is realistic or a scenario we care about, but there is a situation that can hit this path. Because anyone can create custom `MemberInfo` types (e. g. by deriving from `FieldInfo`, they can create a type whose `GetCustomAttributes()` function returns null or a non-`Attribute` array type.

I added a test case that shows this path.

jkotas replied:
Good point. I do not think we care about rare cases where the API violate their contracts. (A similar discussion comes up e.g. for Streams that violate the stream contract.)

jkotas commented:
> is your proposal to return Array.Empty<Attribute>() in that case instead of null?

I think it needs to return Attribute array of the right size.

> Do we still want to use the exception-based control flow?

It would be nice to use non-exception based control flow, same as CoreCLR.

> Also, do you know if there is test coverage that should be updated for this?

We are working to get the existing tests enabled for NativeAOT. No need to add special test coverage.

madelson committed:
Address feedback from https://github.com/dotnet/runtime/pull/65237

madelson committed:
Fixes for https://github.com/dotnet/runtime/pull/65237

Enable open generic attribute type requests for parameters,
break out larger tests into separate cases.

jkotas commented on src/libraries/System.Runtime/tests/System/Attributes.cs:
```diff
@@ -533,6 +536,123 @@ public static void NegTest8()
+        public static void GetCustomAttributesBehaviorMemberInfoReturnsNull()
+        {
+            FieldInfo field = new CustomFieldInfo(null!);
+            if (PlatformDetection.IsMonoRuntime)
```
Should we fix Mono to return null in this corner case as well to avoid this condition?

madelson replied:
@jkotas I'll defer to your decision but I'm not sure it makes sense. Looking through the code it seems like null vs throw in both CoreCLR and Mono are going to be inconsistent depending on the member type and other factors (e. g. whether inherit is true or false). It would take quite a bit more test coverage and probably quite a few changes to get to a consistent state across all the different call flows.

Another option would just be to stop testing this case, since it is so niche so perhaps undefined behavior is fine.

Thoughts?

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:
```diff
@@ -1480,15 +1473,32 @@ private static object CreateCaObject(RuntimeModule module, RuntimeType type, IRu
-            return (object[])Array.CreateInstance(elementType, elementCount);
+            return (object[])Array.CreateInstance(caType, elementCount);
+
+            T[] CreateArray<T>() => elementCount == 0 ? Array.Empty<T>() : new T[elementCount];
```
(Non-static) local methods with captured arguments are inefficient. The compiler will create display type to pass the captured locals around that comes with a lot of ceremony. 

I think I would inline it. The local method is not worth it. Also, I think it may be worth it to check for Attribute and use the faster non-reflection based path for it. Something like:

```
bool useAttributeArray = false;
bool useObjectArray = false;

if (caType == typeof(Attribute))
{
    useAttributeArray = true;
}
else if (caType.IsValueType)
{
    useObjectArray = true;
}
else if (caType.ContainsGenericParameters)
{
    if (caType.IsSubclassOf(typeof(Attribute)))
    {
        useAttributeArray = true;
    }
    else
    {
        useObjectArray = true;
    } 
}

if (useAttributeArray) (elementCount != 0) ? new Attribute[elementCount] :  return Array.Empty<Attribute>();
if (useObjectArray) (elementCount != 0) ? new object[elementCount] : Array.Empty<object>();
return (elementCount != 0) ? (object[])Array.CreateInstance(caType, elementCount) : caType.GetEmptyArray();
```

jkotas reviewed (approved):
LGTM otherwise. Thank you!

jkotas commented:
Looks like Mono is crashing on the newly added tests even though they are disabled:
```
#13 0x0000007f8d2cfbdc in mono_assertion_message_unreachable (file=0x64 <error: Cannot access memory at address 0x64>, line=6) at /__w/1/s/src/mono/mono/eglib/goutput.c:234
#14 0x0000007f8d115120 in custom_attr_class_name_from_method_token (image=0x5589355e20, method_token=167772165, nspace=0x7fc25c2d58, class_name=0x7fc25c2d50, assembly_token=<optimized out>) at /__w/1/s/src/mono/mono/metadata/custom-attrs.c:2441
```

madelson committed:
Additional follow-up from https://github.com/dotnet/runtime/pull/65237

madelson commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeParameterInfo.cs:
```diff
@@ -509,12 +509,12 @@ public override object[] GetCustomAttributes(bool inherit)
                 throw new ArgumentException(SR.Arg_MustBeType, nameof(attributeType));
 
+            if (MdToken.IsNullToken(m_tkParamDef))
+                return CustomAttribute.CreateAttributeArrayHelper(attributeRuntimeType, 0);
```
This change was required to prevent a failure in Microsoft.CSharp.RuntimeBinder.Tests.AccessTests.PublicMethod().

I feel a bit weird exposing the CreateAttributeArrayHelper method as internal, but this is really the exact logic we want here. It makes this branch consistent with other GetCustomAttribute methods which all allow the consumer to cast the array to the requested attribute type.

Moving this after the argument check presumably has some slight perf cost but ultimately feels more correct (and unless we change CreateAttributeArrayHelper to take `Type` instead of `RuntimeType` we need the check to happen first regardless).

jkotas replied:
Ok. Do we need a test to directly verify the behavior of ParamerInfo.GetCustomAttributes for this case?

madelson replied:
Added!

madelson commented:
> Looks like Mono is crashing on the newly added tests even though they are disabled:

@jkotas @buyaa-n From some experimentation, it seems that this is caused by adding the generic assembly attribute. Any thoughts on how best to address this? I tried putting the assembly/module generic attributes behind `#if !MONO` but that did not seem to work. Another option would be to move the module/attribute tests to https://github.com/dotnet/runtime/pull/65237/files#diff-d55f466b22e54fbf3b96b94f21d22c390ae92ff4047896b0e87c6af28e4dcc6f which I believe is CoreCLR-only.

jkotas commented:
I do not think that it makes sense to keep pilling workarounds for Mono bugs.

@SamMonoRT @lambdageek Could you please take a look and suggest fix for the Mono crash that the CI is hitting?

madelson committed:
Further test cleanup for https://github.com/dotnet/runtime/pull/65237

madelson committed:
Attempt to fix Mono test crashes by moving generic attribute testing for assembly and module into CoreCLR-only tests.

madelson committed:
Fix GenericAttributeTests.cs compilation

madelson committed:
Comment out generic attribute assembly tests

madelson commented:
@jkotas @buyaa-n I'm a bit lost interpreting the current crop of errors. Some of them seem to be in cpp files while others are canceled/timed out. Any advice on where to look?

madelson committed:
Trim more tests that fail on Mono

drieseng commented on src/libraries/System.Runtime/tests/System/Attributes.cs:
```diff
@@ -283,27 +283,6 @@ private static void GenericAttributesTestHelper<TGenericParameter>(Func<Type, At
             Assert.Equal(typeof(GenericAttribute<TGenericParameter>[]), closedGenericAttributes.GetType());
         }
-
-        [Fact]
```
Is this just a test, or are you actually removing tests because they fail on Mono?

madelson replied:
Based on the discussion [here](https://github.com/dotnet/runtime/pull/65237#discussion_r808277460) it seemed like these weren't worth it given the failures on Mono. They're all tests I added in earlier commits, not existing tests.

madelson commented:
Many of the remaining errors from the checks are like [this one](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-65237-merge-41fb246fc5b14bc7be/Common.Tests/1/console.a3c88635.log?sv=2019-07-07&se=2022-03-18T23%3A08%3A57Z&sr=c&sp=rl&sig=uvT5addYT11ahJh%2B%2FCA5nT7rolteBqUlkQQBFd9Eo4M%3D)

Example trace:

```
Fatal error. Internal CLR error. (0x80131506)
   at Tests.System.Net.MultiArrayBufferTests.AddSeveralBytesRepeatedlyAndConsumeSeveralBytesRepeatedly_UsingSliceWithLength_Success()
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Span`1<System.Object> ByRef, System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Reflection.MethodBase.Invoke(System.Object, System.Object[])
```

Any thoughts on where to start looking?

jkotas commented:
If the failure is on Windows 7 x86, it is most likely https://github.com/dotnet/runtime/issues/65890 (ie not related to your changes).

madelson committed:
retrigger checks

buyaa-n commented:
> I do not think that it makes sense to keep pilling workarounds for Mono bugs.
> 
> @SamMonoRT @lambdageek Could you please take a look and suggest fix for the Mono crash that the CI is hitting?

It seems this is the only blocker for this PR, @jkotas should we open an issue for this and unblock the PR? Seems @madelson commented out the tests that causing mono failure: https://github.com/dotnet/runtime/blob/86160e1a7175efb328b9b9c874b2990880bedddf/src/tests/reflection/GenericAttribute/GenericAttributeTests.cs#L25-L30

jkotas commented:
> this is the only blocker for this PR, @jkotas should we open an issue for this and unblock the PR?

Fine with me.

buyaa-n commented on src/tests/reflection/GenericAttribute/GenericAttributeTests.cs:
```diff
@@ -22,6 +22,13 @@ static int Main(string[] args)
+        // Assert(CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<long>)).GetEnumerator().MoveNext());
+        // Assert(CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<long>)).GetEnumerator().MoveNext());
+        // Assert(!CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());
+        // Assert(!CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());	
```
Should we include these tests within the above mulitilne comment? Seems related to me

madelson replied:
I wasn't sure if it was related and wanted to ask about that. It's possible that these tests could be brought back.

Do you know what the preferred approach is for iterating on this particular test locally? [https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/testing.md#build-individual-test](this page) suggests that I should run 

> dotnet.cmd build c:\dev\runtime\src\tests\reflection\GenericAttribute\GenericAttributeMetadata.ilproj -c Release

but that just builds a dll and doesn't seem to actually invoke this main method.

buyaa-n replied:
Sorry idk how to run them locally, @davidwrighton @jkotas might help with that

jkotas replied:
The instructions for running these tests are at https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/windows-test-instructions.md

buyaa-n commented:
> > this is the only blocker for this PR, @jkotas should we open an issue for this and unblock the PR?
> 
> Fine with me.

Thanks, looks there is an open issue for enabling these tests https://github.com/dotnet/runtime/issues/56880

buyaa-n reviewed (approved):
Thanks for working on this @madelson!

buyaa-n commented:
An [issue created](https://github.com/dotnet/runtime/issues/66168) for mono failure, CI failure looks related to [this issue](https://github.com/dotnet/runtime/issues/66083) and unrelated to this change => merging.
